core/dispatcher/src/whisk/core/container/ContainerPool.scala

```diff
@@ -295,6 +309,34 @@ class ContainerPool(
     private def makeContainerName(action: WhiskAction): String =
         ContainerCounter.containerName(invokerInstance.toString(), action.fullyQualifiedName)
 
+    // A background thread that re-populates the container pool with fresh (un-instantiated) nodejs containers.
+    private val warmupThread = new Thread {
+        override def run {
+            while (true) {
+                val tid = TransactionId(0)
+                if (getNumberOfIdleContainers(warmNodejsKey)(tid) < WARM_NODEJS_CONTAINERS) {
+                    makeWarmNodejsContainer()(tid)
+                }
+                Thread.sleep(100)
+            }
+        }
+    }
+
```

Comment:
1. "keyMap.get(key) map "  sure - i remain mildly nervous about how good higher order inlining is in scala so i tend towards the lest lambda-y stuff inside sync blocks
2. "if don't care about transid, use TransactionId.dontcare"  will do
3. "should this be consolidated with NodeJSExec.image?"   What do you mean?
4. "so there will one one warm up thread..."?  No.  Just one thread for all warmup but i want to focus on nodejs first (not least to minimize perf risk).